File: app/src/main/java/li/klass/fhem/appwidget/view/widget/base/otherWidgets/OtherWidgetsAdapter.java
package li.klass.fhem.appwidget.view.widget.base.otherWidgets;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.google.common.collect.Lists;

import org.apache.commons.lang3.tuple.Pair;

import java.util.Comparator;

import li.klass.fhem.R;
import li.klass.fhem.adapter.ListDataAdapter;
import li.klass.fhem.appwidget.view.WidgetType;

public class OtherWidgetsAdapter extends ListDataAdapter<Pair<WidgetType, String>> {
    public OtherWidgetsAdapter(Context context) {
        super(context, R.layout.simple_list_item, Lists.<Pair<WidgetType, String>>newArrayList(),
                new Comparator<Pair<WidgetType, String>>() {
            @Override
            public int compare(Pair<WidgetType, String> pair1, Pair<WidgetType, String> pair2) {
                return pair1.getValue().compareTo(pair2.getValue());
            }
        });
    }

    @Override
    @SuppressWarnings("unchecked")
    public View getView(int position, View view, ViewGroup parent) {
        Pair<WidgetType, String> item = (Pair<WidgetType, String>) getItem(position);

        if (view == null) {
            view = inflater.inflate(resource, null);
        }

        TextView textView = (TextView) view.findViewById(R.id.text);
        textView.setText(item.getValue());

        view.setTag(item.getKey());

        return view;
    }
}

File: app/src/main/java/li/klass/fhem/appwidget/ui/widget/base/otherWidgets/OtherWidgetsAdapter.kt
package li.klass.fhem.appwidget.ui.widget.base.otherWidgets

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import kotlinx.android.synthetic.main.simple_list_item.view.*
import li.klass.fhem.R
import li.klass.fhem.appwidget.ui.widget.base.AppWidgetView

class OtherWidgetsAdapter(elements: List<AppWidgetView>, val widgetClickedCallback: OtherWidgetsFragment.OnWidgetClickedCallback)
    : RecyclerView.Adapter<OtherWidgetsAdapter.ViewHolder>() {
    private val sortedElements = elements.sortedBy { it.getWidgetName() }

    override fun getItemCount() = sortedElements.size
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val layout = LayoutInflater.from(parent.context).inflate(R.layout.one_line_card, parent, false)
        return ViewHolder(layout, widgetClickedCallback)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(sortedElements[position])
    }

    class ViewHolder(val view: View, val widgetClickedCallback: OtherWidgetsFragment.OnWidgetClickedCallback) : RecyclerView.ViewHolder(view) {
        fun bind(appWidgetView: AppWidgetView) {
            view.text.text = view.context.getString(appWidgetView.getWidgetName())
            view.setOnClickListener { widgetClickedCallback.onWidgetClicked(appWidgetView) }
        }
    }
}
